Bugfix/price change subscribe

kossme opened the pull request:
Для проверки: 
1. Дропнуть БД и создать заново
2. Зарегаться как новый юзер 
3. Подтвердить регистрацию 
4. Войти и подписаться на изменение цены любого товара (или нескольких). в БД  в таблице Users значение confirm_receive_email поменяется на Requested 
5. Зайти в почту: должно прийти письмо-запрос подтверждения рассылок. Ответить на него (неважно чем)
6. В течении 30 сек. в БД поменяется статус на Confirmed
7. Зайти под менеджером и поменять цену в меньшую сторону на товар, на который подписан тот юзер (или товары).
8. Проверить почту с оповещением о снижении цены

kossme committed:
add new Column in User - confirmationReceiveEmails.
new method sendConfirmationSubscribeLetter.
New message template in DataInitializer.

kossme committed:
Добавлен новый столбец в таблицу User - confirmationReceiveEmails, может принимать 3 значения-статуса относительно согласия рассылки (enum): No_Actions, Requested, Confirmed. При нажатии юзером на "Подписаться", направляется письмо с просьбой дать согласие на подписку, направив любое ответное письмо. Это статус Requested. Пока юзер не даст согласие, такое письмо приходит ему каждый раз при нажатии на "Подписаться".
При этом адрес юзера заносится в список рассылок, но сама рассылка не приходит (иначе забанят за спам).

После получения ответного описьма от юзера, статус переходит в Confirmed.
После чего он начинает получать рассылки при снижении цен на подписанные товары.

add new Column in User - confirmationReceiveEmails.
new method sendConfirmationSubscribeLetter in UserService.
New message template in DataInitializer.
New ReceiveEmailSubscribeConfirmation and impl

kossme committed:
Merge branch 'dev' into bugfix/priceChangeSubscribe

kossme commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Проблема с отправкой почты: подписка на изменение цены товара](https://trello.com/c/4d8zcAHg/210-%D0%BF%D1%80%D0%BE%D0%B1%D0%BB%D0%B5%D0%BC%D0%B0-%D1%81-%D0%BE%D1%82%D0%BF%D1%80%D0%B0%D0%B2%D0%BA%D0%BE%D0%B9-%D0%BF%D0%BE%D1%87%D1%82%D1%8B-%D0%BF%D0%BE%D0%B4%D0%BF%D0%B8%D1%81%D0%BA%D0%B0-%D0%BD%D0%B0-%D0%B8%D0%B7%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-%D1%86%D0%B5%D0%BD%D1%8B-%D1%82%D0%BE%D0%B2%D0%B0%D1%80%D0%B0)

kossme committed:
Add JavaDoc

Добавлен новый столбец в таблицу User - confirmationReceiveEmails, может принимать 3 значения-статуса относительно согласия рассылки (enum): No_Actions, Requested, Confirmed. При нажатии юзером на "Подписаться", направляется письмо с просьбой дать согласие на подписку, направив любое ответное письмо. Это статус Requested. Пока юзер не даст согласие, такое письмо приходит ему каждый раз при нажатии на "Подписаться".
При этом адрес юзера заносится в список рассылок, но сама рассылка не приходит (иначе забанят за спам).

После получения ответного описьма от юзера, статус переходит в Confirmed.
После чего он начинает получать рассылки при снижении цен на подписанные товары.

add new Column in User - confirmationReceiveEmails.
new method sendConfirmationSubscribeLetter in UserService.
New message template in DataInitializer.
New ReceiveEmailSubscribeConfirmation and impl

kossme committed:
Merge branch 'dev' into bugfix/priceChangeSubscribe

# Conflicts:
#	src/main/java/com/jm/online_store/service/impl/UserServiceImpl.java

kossme committed:
Add JavaDoc
Исправлены ошибки

Добавлен новый столбец в таблицу User - confirmationReceiveEmails, может принимать 3 значения-статуса относительно согласия рассылки (enum): No_Actions, Requested, Confirmed. При нажатии юзером на "Подписаться", направляется письмо с просьбой дать согласие на подписку, направив любое ответное письмо. Это статус Requested. Пока юзер не даст согласие, такое письмо приходит ему каждый раз при нажатии на "Подписаться".
При этом адрес юзера заносится в список рассылок, но сама рассылка не приходит (иначе забанят за спам).

После получения ответного описьма от юзера, статус переходит в Confirmed.
После чего он начинает получать рассылки при снижении цен на подписанные товары.

add new Column in User - confirmationReceiveEmails.
new method sendConfirmationSubscribeLetter in UserService.
New message template in DataInitializer.
New ReceiveEmailSubscribeConfirmation and impl

tediik commented on src/main/java/com/jm/online_store/service/impl/UserServiceImpl.java:
```diff
@@ -573,4 +578,34 @@ public User getUserByToken(String token) {
+        if (user.isPresent()) {
+            User userFromDB = user.get();
+            userFromDB.setConfirmReceiveEmail(User.ConfirmReceiveEmail.REQUESTED);
+            System.out.println(userFromDB.getConfirmReceiveEmail());
```
это еще зачем?

tediik commented on src/main/java/com/jm/online_store/service/impl/UserServiceImpl.java:
```diff
@@ -573,4 +578,34 @@ public User getUserByToken(String token) {
+                .getTextValue();
+        String messageBody;
+        String usersName = "Покупатель";
+        Optional<User> user = findByEmail(email);
+        if (user.isPresent()) {
+            User userFromDB = user.get();
```
вот эти три строки по сути заменяет метод ifPresent у Optional, перепиши, потренируешься заодно писать красивый правильный код

tediik commented on src/main/java/com/jm/online_store/service/impl/ReceiveEmailSubscribeConfirmationImpl.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.Optional;
+import java.util.Properties;
+
+/**
+ * Метод проверяет почту на предмет подтверждения юзерами подписки на получение писем об изменении цены
+ * (в ответ на запрошенное подтверждение).
+ * Сверяет адресаты полученных непрочитанных писем с наличием их в БД и статусом ConfirmReceiveEmail.REQUESTED, при
+ * совпадении - меняет статус на ConfirmReceiveEmail.CONFIRMED, после чего пользователь будет получать рассылку.
+ * (Подтверждение запрашивается при нажатии "Подписаться на изменение цены" каждый раз, пока не будет получено согласие)
+ */
```
отлично!

tediik commented on src/main/java/com/jm/online_store/model/User.java:
```diff
@@ -228,6 +234,16 @@ public boolean isEnabled() {
         WOMAN
     }
 
+    /**
+     * Обозначает статус согласия юзера на получение рассылки (об изменении цены).
+     * NO_ACTIONS - нет согласия, REQUESTED - согласие запрошено (в письме), CONFIRMED - согласие получено.
+     */
+    public enum ConfirmReceiveEmail {
+        NO_ACTIONS,
+        REQUESTED,
+        CONFIRMED
+    }
+
```
глянь, вроде перечисления выносили в отдельный пакет же?

tediik commented on src/main/java/com/jm/online_store/config/DataInitializer.java:
```diff
@@ -1266,6 +1273,12 @@ public void commonSettingsInit() {
+                .settingName("subscribe_confirmation_template")
+                .textValue("<p>Уважаемый @@user@@, для подтверждении подписки о снижении цены</p>" +
+                        "<p>Пожалуйста, ответьте на это письмо с любым текстом</p>" +
+                        "<p>С Уважением</p><p>Online-store.ru</p>")
```
посмотри, у нас сейчас в application.yml есть константа вроде production-url - используй ее, это всяко лучше, чем потом по коду искать, кто где название вписал хардкорно

tediik commented on src/main/java/com/jm/online_store/model/Comment.java:
```diff
@@ -65,8 +65,7 @@
 
-    @Column(nullable = false)
-    private boolean deletedHasKids = false;
+     private boolean deletedHasKids = false;
```
не понял как это изменение относится к твоей задаче, но если считаешь, что так нужно - ОК

kossme committed:
Add JavaDoc
Исправлены ошибки

Добавлен новый столбец в таблицу User - confirmationReceiveEmails, может принимать 3 значения-статуса относительно согласия рассылки (enum): No_Actions, Requested, Confirmed. При нажатии юзером на "Подписаться", направляется письмо с просьбой дать согласие на подписку, направив любое ответное письмо. Это статус Requested. Пока юзер не даст согласие, такое письмо приходит ему каждый раз при нажатии на "Подписаться".
При этом адрес юзера заносится в список рассылок, но сама рассылка не приходит (иначе забанят за спам).

После получения ответного описьма от юзера, статус переходит в Confirmed.
После чего он начинает получать рассылки при снижении цен на подписанные товары.

add new Column in User - confirmationReceiveEmails.
new method sendConfirmationSubscribeLetter in UserService.
New message template in DataInitializer.
New ReceiveEmailSubscribeConfirmation and impl

tediik reviewed (approved):
Отлично
